ashura/engine/animation.h

```diff
@@ -2,25 +2,620 @@
 #include "ashura/std/types.h"
+#include <concepts>
+#include <deque>
+#include <functional>
+#include <map>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
 
 namespace ash
 {
 
 /// [ ] https://create.roblox.com/docs/ui/animation#style
 /// [ ] Procedural Animation https://www.youtube.com/watch?v=qlfh_rv6khY
 
-// [ ] Keyframes, for time interval x, move from a to b
+// [x] Keyframes, for time interval x, move from a to b
 // [ ] keyframe blending
-// [ ] play
-// [ ] reverse
-// [ ] cancel
+// [x] play
+// [x] reverse
+// [x] cancel
 // [ ] frame-rate customization
-// [ ] move back from point
-// [ ] loop n or forever
+// [x] move back from point
+// [x] loop n or forever
 // [ ] stop after timepoint
 // [ ] alternate
-// [ ] pause
-// [ ] resume in direction
-// [ ] reset animation state to beginning
+// [x] pause
+// [x] resume in direction
+// [x] reset animation state to beginning
+
+// TODO: Support ash Span type instead of using std::string
+
+// Animatable trait applicable to complex types like Color, Vec2, Vec3 etc
+template <typename T>
+concept Animatable = requires(T a, T b, f32 t) {
+  { lerp(a, b, t) } -> std::convertible_to<T>;
+};
+
+namespace animation
+{
+// Lerp implementations for different types
+// Basic arithmetic types
+template <typename T>
+T lerp(const T &start, const T &end, f32 t)
+  requires std::is_arithmetic_v<T>
+{
+  return ash::lerp<T>(start, end, t);
+}
+}        // namespace animation
+
+template <Animatable T>
+struct Animation;
+template <typename T>
+struct Keyframe;
+template <typename T>
+struct Timeline;
+
+struct AnimationTime
+{
+  using Clock     = std::chrono::steady_clock;
+  using TimePoint = Clock::time_point;
+  using Duration  = std::chrono::duration<float>;
+
+  static f32 get_delta(const TimePoint &current, const TimePoint &previous)
+  {
+    return std::chrono::duration_cast<Duration>(current - previous).count();
+  }
+};
+
+enum class CurveType
+{
+  Linear,
+  EaseIn,
+  EaseOut,
+  EaseInOut,
+  Custom
+};
+
+struct AnimationConfig
+{
+  f32         duration = 1.0f;
+  f32         delay    = 0.0f;
+  bool        loop     = false;
+  CurveType   easing   = CurveType::Linear;
+  std::string label    = "default";
+};
+
+struct AnimationState
+{
+  bool is_playing   = false;
+  f32  current_time = 0.0f;
+  f32  delay        = 0.0f;
+};
+
+struct TimelineOptions
+{
+  f32         delay    = 0.0f;
+  bool        autoplay = true;
+  bool        loop     = false;
+  std::string label    = "default";
+};
+
+enum class GridDirection
+{
+  RowMajor,          // Left to right, top to bottom
+  ColumnMajor        // Top to bottom, left to right
+};
+
+struct StaggerOptions
+{
+  f32 start = 0.0f;
+  f32 from  = 0.0f;
+  struct Grid
+  {
+    i32           rows      = 1;
+    i32           cols      = 1;
+    GridDirection direction = GridDirection::RowMajor;
+  } grid;
+  std::function<f32(i32 row, i32 col, i32 index, i32 total)> easing = nullptr;
```

Comment:
easing functions typically don't carry much of data beyond say 4 floats? I feel we could optimize for this and use for example:

```cpp
struct Easing{
typedef f32(*Function)(f32 [4]);
f32 params[4] = {};
Function easing = linear; // <- note, we use reasonable defaults where possible
};
```